wire-ios-data-model/Source/Public/GetUserClientFingerprintUseCase.swift

```diff
@@ -0,0 +1,129 @@
+////
+// Wire
+// Copyright (C) 2023 Wire Swiss GmbH
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see http://www.gnu.org/licenses/.
+//
+
+import Foundation
+
+public class GetUserClientFingerprintUseCase {
+
+    let proteusProvider: ProteusProviding
+    let managedObjectContext: NSManagedObjectContext
+
+    public static func create(for managedObjectContext: NSManagedObjectContext) -> GetUserClientFingerprintUseCase {
+        let proteusProvider = ProteusProvider(context: managedObjectContext)
+
+        return GetUserClientFingerprintUseCase(proteusProvider: proteusProvider, managedObjectContext: managedObjectContext)
+    }
+
+    init(proteusProvider: ProteusProviding, managedObjectContext: NSManagedObjectContext) {
+        self.proteusProvider = proteusProvider
+        self.managedObjectContext = managedObjectContext
+    }
+
+    public func invoke(userClient: UserClient) async -> Data? {
+        let objectId = userClient.objectID
+
+        let isSelfClient: Bool = managedObjectContext.performAndWait {
+            guard let existingUserId = try? managedObjectContext.existingObject(with: objectId) as? UserClient else {
+                return false
+            }
+            return existingUserId.isSelfClient()
+        }
+
+        if isSelfClient {
+            return await localFingerprint()
+        } else {
+            return await fetchRemoteFingerprint(for: userClient)
+        }
+    }
+
+    func localFingerprint() async -> Data? {
+        var canPerform: Bool = false
+
+        managedObjectContext.performAndWait {
+            canPerform = proteusProvider.canPerform
+        }
+        guard canPerform else {
+            return nil
```

Comment:
Would it be helpful to throw an error or we don't really care what went wrong?

Reply:
added a log and kept it non-throwing for now